tests/unittests/OperatorTest.cpp

```diff
@@ -1389,11 +1388,71 @@ void checkFloat16Convolution(ExecutionEngine &EE, unsigned convDepth) {
   }
 }
 
-TEST_P(InterpOnly, FP16ConvolutionDepth10) { checkFloat16Convolution(EE_, 10); }
+TEST_P(InterpOnly, FP16ConvolutionDepth10) {
+  checkFloat16Convolution(EE_, F_, 10);
+}
+
+TEST_P(InterpOnly, FP16ConvolutionDepth8) {
+  checkFloat16Convolution(EE_, F_, 8);
+}
+
+void checkFloat16ConvolutionWithCast(ExecutionEngine &EE, Function *F,
+                                     unsigned convDepth) {
+  // In this test we generate a half precision floating-point based
+  // convolution graph.
+  // Compile, run it.
+  // Then, convert the graph into single precision.
+  // Compile, run it.
+  // Finally, we check that the output are similar.
+
+  auto &mod = EE.getModule();
+  Context ctx;
+
+  auto *inputFloat16 = mod.createPlaceholder(
+      ElemKind::Float16Ty, {1, 10, 10, 3}, "in_float16", false);
+  ctx.allocate(inputFloat16);
+  auto *convFloat16 =
+      F->createConv(ctx, "convFloat16", inputFloat16, convDepth, 5, 1, 0, 1);
+
+  Tensor *inputFloat16Tensor = ctx.get(inputFloat16);
+  auto inputFloat16Handle = inputFloat16Tensor->getHandle<float16_t>();
+  inputFloat16Handle.randomize(-1.0, 1.0, mod.getPRNG());
 
-TEST_P(InterpOnly, FP16ConvolutionDepth8) { checkFloat16Convolution(EE_, 8); }
+  SaveNode *saveFloat16 = F->createSave(ctx, "saveFloat16", convFloat16);
 
-void checkIntConvolution(ExecutionEngine &EE, unsigned convDepth,
+  Tensor *out = ctx.allocate(llvm::cast<Placeholder>(saveFloat16->getOutput()));
+
+  EE.compile(CompilationMode::Infer, F, ctx);
+
+  EE.run();
+
+  // Save the value of the fp16 run.
+  Tensor float16OutTensor = out->clone();
+
+  // Convert the graph to single precision.
+  F->mutateNodesType(ElemKind::Float16Ty, ElemKind::FloatTy, &ctx);
```

Comment:
I don't think that this method makes sense to me. The quantizer clones the function and creates a new function that operates on in integers. The two methods should be compatible. There is no reason to have two different kinds of mechanisms, and especially not being a member of the function.

Reply:
I agree, and I’d argue the quantizer does it wrong. Indeed each type we add a new node, we have to go modify the quantization and create the node manually with the right type.

I agree we can do this on the site, but I equally think it could be part à the function API.

I don’t have a strong opinion either way. If you do I am happy to move the functionality to the side.

Reply:
Regardless of which design is correct they should be consistent. At the moment your API handles the inputs/outputs and the processing of the network itself in a completely different way. 

 I like the design where we clone the function before modifying it because it allows to keep the original function and use it for testing the outputs. 

If you think that the implementation of the quantizer can be changed then let's do it. If you think that the API of the quantizer can be changed then let's also do that. But let's make them consistent. This includes the loader and the compiler.